### PR TITLE
[WIP] Add maturity file and clarify requirements

### PR DESCRIPTION
**Pull request by brian-avery**
I've heard feedback that the current list of requirements can be confusing. This PR attempts to address the changes made to the experimental list and extend them to alpha/beta/stable. It also works towards addressing how approvals may be tracked.

@linsun @sdake @ericvn @ostromart @nmittler can you please take a look at this and let me know your thought based on what you're seeing?

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 1]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
+
+[//]: # (A design doc is mandatory for graduation to beta. The design doc must be
+approved by relevant work group leads.)
+
 - [ ] Feature coverage and test plans written and approved.
+[//]: # (Feature coverage is required for graduation to beta. More details can be found here: https://github.com/istio/istio/blob/8af80bdf91684d8e517e9e5d57f2c3d8480b11a2/pkg/test/framework/features/README.md)
 
 **Docs** 
 
 - [ ] Documentation on istio.io includes performance expectations; may have caveats.
```
Comment (ostromart) [inline .github/ISSUE_TEMPLATE/beta_feature.md:63]: examples? how to know if relevant for feature?

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 2]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
+approved by relevant work group leads.)
+
 - [ ] Feature coverage and test plans written and approved.
+[//]: # (Feature coverage is required for graduation to beta. More details can be found here: https://github.com/istio/istio/blob/8af80bdf91684d8e517e9e5d57f2c3d8480b11a2/pkg/test/framework/features/README.md)
 
 **Docs** 
 
 - [ ] Documentation on istio.io includes performance expectations; may have caveats. 
 - [ ] Documentation on istio.io includes samples/tutorials. 
 - [ ] Documentation on istio.io includes appropriate glossary entries.
```
Comment (ostromart) [inline .github/ISSUE_TEMPLATE/beta_feature.md:65]: same comment as above

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 3]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
+
 - [ ] Feature coverage and test plans written and approved.
+[//]: # (Feature coverage is required for graduation to beta. More details can be found here: https://github.com/istio/istio/blob/8af80bdf91684d8e517e9e5d57f2c3d8480b11a2/pkg/test/framework/features/README.md)
 
 **Docs** 
 
 - [ ] Documentation on istio.io includes performance expectations; may have caveats. 
 - [ ] Documentation on istio.io includes samples/tutorials. 
 - [ ] Documentation on istio.io includes appropriate glossary entries. 
 - [ ] All new documentation containing user actions includes istio.io tests.
```
Comment (ostromart) [inline .github/ISSUE_TEMPLATE/beta_feature.md:66]: why is it specifically new documentation here? shouldn't it be all?

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 4]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
+
+[//]: # (Documentation tests are required in order to be able to confirm that
+the documentation is correct across Istio releases. Details on writing Istio
+documentation tests can be found here: https://github.com/istio/istio.io/tree/master/tests)
+
 - [ ] Release notes have been added. 
 - [ ] Upgrade notes have been added.
```
Comment (ostromart) [inline .github/ISSUE_TEMPLATE/beta_feature.md:74]: below: how to determine if fixed issues have regression tests? manual scrub of all past issues?

File: .github/ISSUE_TEMPLATE/alpha_feature.md [hunk 5]
```diff
@@ -3,18 +3,44 @@ name: Alpha Feature Requirements
 about: Requirements for promoting a feature to alpha
 ---
 
-# Alpha Feature Requirements
+[//]: # (The syntax preceeding this line is a comment marker used to help guide the issue author with submission of an issue)
+[//]: # (to github. Unlike HTML comments commonly used throughout istio.io documentation, this comment will not be rendered)
+[//]: # (by github. Comments must be separated by carriage return preceding and concluding the text and be a single line.)
 
-This page lists the requirements for promoting a feature to alpha. Please check off and document the steps as they are completed.
+[//]: # (This page lists the requirements for an experimental feature. Completion of this template enables Istio work groups)
```
Comment (sdake) [inline .github/ISSUE_TEMPLATE/alpha_feature.md:10]: experimental->alpha.

File: .github/ISSUE_TEMPLATE/alpha_feature.md [hunk 6]
```diff
@@ -3,18 +3,44 @@ name: Alpha Feature Requirements
...
+[//]: # (to collect information on potential new functionality. This template should be completed before users are exposed to)
+[//]: # (any new experimental feature. Please complete this template during development.)
```
Comment (sdake) [inline .github/ISSUE_TEMPLATE/alpha_feature.md:12]: Lines 10-12 should describe the goals of the template for alpha.

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 7]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
+[//]: # (The syntax preceeding this line is a comment marker used to help guide the issue author with submission of an issue)
+[//]: # (to github. Unlike HTML comments commonly used throughout istio.io documentation, this comment will not be rendered)
+[//]: # (by github. Comments must be separated by carriage return preceding and concluding the text and be a single line.)
 
-This page lists the requirements for promoting a feature to beta. Promotion to beta must meet all requirements for promotion to alpha. Please check off and document the steps as they are completed.
+[//]: # (This page lists the requirements for an experimental feature. Completion of this template enables Istio work groups)
+[//]: # (to collect information on potential new functionality. This template should be completed before users are exposed to)
+[//]: # (any new experimental feature. Please complete this template during development.)
+
+[//]: # (The feature implementation section must be completed before submission of the issue.)
```
Comment (sdake) [inline .github/ISSUE_TEMPLATE/beta_feature.md:14]: 10-14 should be modified for beta in some way.

File: .github/ISSUE_TEMPLATE/beta_feature.md [hunk 8]
```diff
@@ -3,35 +3,72 @@ name: Beta Feature Requirements
...
-**Feature Name:** 
+**RFC:**
 
-**Design Doc:**
+[//]: # (This outlines the ideas for a feature and is used to collect early feedback. See the RFC template here: https://docs.google.com/document/d/1ewJoCcw5-04crH-M0xw4zFxz1cfwVCPnNyW4K3m4Yyc/template/preview)
 
-**Alpha Checklist:**
+**Design doc:**
+
+[//]: # (This is used for discussion of design and testing for a feature. See the design doc template here:https://docs.google.com/document/d/16FLQK8uhhic1ovKnnOG3OXJjFKs2aHnSmbximidpKwM/template/preview)
```
Comment (sdake) [inline .github/ISSUE_TEMPLATE/beta_feature.md:38]: this type of comment may confuse the rendering - I can't remember if I was able o render HTTP correctly vs not have it rendered when the issue was created. Can you verify?

File: .github/ISSUE_TEMPLATE/stable_feature.md [hunk 9]
```diff
@@ -3,18 +3,47 @@ name: Stable Feature Requirements
...
+[//]: # (to github. Unlike HTML comments commonly used throughout istio.io documentation, this comment will not be rendered)
+[//]: # (by github. Comments must be separated by carriage return preceding and concluding the text and be a single line.)
 
-This page lists the requirements for promoting a feature to stable. Promotion to stable must meet all requirements of promotion to beta. Please check off and document the steps as they are completed.
+[//]: # (This page lists the requirements for an experimental feature. Completion of this template enables Istio work groups)
+[//]: # (to collect information on potential new functionality. This template should be completed before users are exposed to)
+[//]: # (any new experimental feature. Please complete this template during development.)
 
-**Feature Name:** 
+[//]: # (The feature implementation section must be completed before submission of the issue.)
```
Comment (sdake) [inline .github/ISSUE_TEMPLATE/stable_feature.md:14]: 10-14 - same comment.

**Review by sdake — COMMENTED** [review]
looks good - thanks @brian-avery for taking this on. I had planned to complete this post branch as 3 separate PRs, but I'm glad your on it 👍 

I think the beta bar is a little too high and may result in features remaining in alpha for very extended periods.

IMO the progression should be linear, not logarithmic.

Cheers,
-steve

**Comment by sdake** [comment]
@linsun For some reason I have the ability to merge PRs in this repo:
Screen Shot 2020-11-08 at 3 49 47 PM

@brian-avery this PR seems blocked waiting on addressing review comments. Did you want me to take it over 1 by 1, or do you plan to address the comments?

@linsun is the goal of alpha->beta to be logamarthic, rather than linear? I think linear would be better, or features will forever be stuck in alpha.

Cheers,
-steve

**Comment by zerobfd** [comment]
I'm not sure linear/logarithmic is the right way to think about it.

My interpretation of [Istio's feature lifecycle](https://github.com/istio/community/blob/master/FEATURE-LIFECYCLE.md) is that the relationship of alpha to beta isn't the same as beta to stable. From alpha to beta is progression in the sense that the feature is still being worked on and tested. Marking something as "beta" means that it's reasonably expected to be done. In the ideal case, beta to stable includes zero changes, it just means that the feature has been out in the world long enough with no bugs and we have confidence that it's safe.

In our feature lifecycle doc in particular, the fact that we say that beta features can be shipped on-by-default implies this. This also lines up with other projects like [Kubernetes](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) and [API improvement proposals](https://google.aip.dev/181).

It does mean that features may stay in alpha for a long period of time, which is discouraging to a lot of people. I think that the answer to that may be some clarification that alpha -> beta -> stable isn't a good indicator of progress of a feature. (Maybe checking off boxes on the beta checklist would be a better way to measure that?)

**Pull request closed without merging**